### ISSUE TITLE
How to show hidden dotfiles without undoing sort?

### ISSUE THREAD

**Issue by airblade** [justinmk/vim-dirvish#46]
I am trying to customise the directory listing like this:

- Directories are listed before files.
- Directories and files are each sorted alphabetically.
- Dotfiles are hidden by default.

This is what I've got:

```viml
augroup my_dirvish_events
  autocmd!

  " Sort: folders at top, alphabetical, case-insensitive.
  autocmd FileType dirvish sort ir /^.*[^\/]$/

  " Hide dotfiles by default.
  " To unhide, type u to undo.  To hide again, reload with R.
  autocmd FileType dirvish silent keeppatterns g|\v/\.[^\/]+/?$|d
augroup END
```

It works until I try to show the dotfiles with `u`: instead of only unhiding them, it also un-does the sort.  I'm not sure if/why vim is treating the `sort` and `global` commands as a single undo-able item.

I'm sure there's a simple solution but I can't seem to find it...

Thanks!

**Comment by justinmk** [comment]
If you comment out the `setlocal undolevels` lines in `s:buf_render`, does it work?

**Comment by airblade** [comment]
Unfortunately not.

L282 commented out | L287 commented out | Result of `u`
---- | ---- | ----
N | N | both sort and global commands undone
N | Y | no change: "already at oldest change"
Y | N | both sort and global commands undone
Y | Y | buffer emptied

**Comment by justinmk** [comment]
Hmm, the "buffer emptied" result makes me think this is how Vim normally treats `buftype=nofile`.

With L282/L287 commented-out, try removing `buftype=nofile` from this line:

https://github.com/justinmk/vim-dirvish/blob/c76e3d3eb607868490d8ad5ceb162a35b29fc062/autoload/dirvish.vim#L106

**Comment by airblade** [comment]
The buffer is still emptied.  If I then try to quit Vim it warns me about the buffer being modified and unsaved (as you'd expect).

**Comment by justinmk** [comment]
I guess Vim is treating the `autocmd FileType dirvish` group as a single script context. Possibly an "undo point" could be forced via feedkeys() hackery..

**Comment by justinmk** [comment]
You may find https://github.com/justinmk/vim-dirvish/pull/31 useful.

**Comment by airblade** [comment]
Thanks for the link to #31.

Vim's certainly behaving like it treats the filetype group as a single script context.

I tried to break the undo sequence with `autocmd FileType dirvish execute "normal! i<C-G>u"` between my sort and global commands but it didn't make any difference.  I tried some `feedkeys()` incantations but I really don't know how to use that, and I couldn't get it to work.  I also tried moving the global command into a function and calling the function from the autocmd...but it didn't help.

It seems like there isn't an easy way to make the commands separately undoable after all, in which case I'm fine with having the dotfiles listed.  I'll ditch the `:g/...d` and keep the `sort`.

Thanks for helping me investigate this.
